wpengine/atlas-content-modeler
# MTKA-1332: model/taxonomy label check for collisions

[Edited by Nick on 11 Jan 2021]

## Description

Ensures that models and taxonomies can not be created or updated if:

- An ACM singular or plural model label collides with another model or taxonomy label. (GraphQL registers them at the top level, so labels must be globally unique.)
- The singular or plural label of the model or taxonomy would collide with a reserved top-level GraphQL “field”, such as 'post', 'posts', 'category', categories', 'plugin', 'plugins' etc. 

<img width="425" alt="Screenshot 2022-01-11 at 14 10 19" src="https://user-images.githubusercontent.com/647669/148948400-ec5affda-3322-447b-be09-f1502c4b5d4e.png">

<img width="429" alt="Screenshot 2022-01-11 at 14 11 19" src="https://user-images.githubusercontent.com/647669/148948537-f2f82ed1-6de9-447e-a291-021dba28da25.png">

Also checks models that are bulk imported via the Import Tool. If any model in the group contains a bad label the full import will be rejected so that the user can correct the file and re-import. (I chose to reject the import in full so the user doesn't have to trim any models from their import file that were correctly imported the first time.)

<img width="1043" alt="Screenshot 2022-01-11 at 13 58 42" src="https://user-images.githubusercontent.com/647669/148947793-66c6277c-b9dc-46f5-9764-2231de3223c0.png">

https://wpengine.atlassian.net/browse/MTKA-1332

## Automated Tests
Includes:
- Unit tests for the REST endpoint updates.
- End-to-end tests to confirm user feedback is provided if an illegal label is used.

## Manual Testing
You should no longer be able to create a model or taxonomy with the same single or plural label as another model or taxonomy. (This requires that WPGraphQL is active.)

You should not be able to create models or taxonomies with a singular or plural label that uses a reserved term such as "post", "posts", "plugin" or "plugins". (This part should work even if WPGraphQL is inactive.)
